rfcs/0075-declarative-wrappers.md

```diff
@@ -0,0 +1,372 @@
+automatic as possible. A derivation that needs some environment variables in
+order to work will get these environment variables set in the wrapper by
+`mkDerivation`'s standard `fixupPhase`.  As a start, we'll (me probably) will
+introduce a new `makeWrapperAuto` setup hook that will take care of this in the
+way described as follows.
+
+As a start, we'll need to think about all packages in Nixpkgs that using them
+requires some environment variables to be set. Every such package will put in
+`$dev/nix-support/wrappers.json` a list of environment variables that are
+"linked" to this package. "linked" means that using this package requires these
+environment variables to be set in runtime.
+
+`makeWrapperAuto` will traverse all `buildInputs` and `propagatedBuildInputs`
+of a derivation, and look for a `wrappers.json` file in these inputs. It will
+collect all the environment variables that need to be set in the resulting
+executables, by merging all of the values of the environment variables in all
+of the inputs' `wrappers.json` files. `wrappers.json` might look like this for
+a given package:
+
+```json
+{
+  "GI_TYPELIB_PATH": [
+    "/nix/...gobject-introspection.../...",
+    "/nix/...librsvg.../...",
+  ],
+  "GIO_EXTRA_MODULES": [
+    "/nix/...dconf.../lib/gio/modules"
+  ],
+  "XDG_DATA_DIRS": [
+    "/nix/...gtk+3.../...",
+    "/nix/...gsettings-desktop-schemas.../..."
+  ],
+  "GDK_PIXBUF_MODULE_FILE": "/nix/...librsvg.../lib/gdk.../loaders.cache",
+}
+```
+
+The information found inside an input's `wrappers.json` will specify the
+wrapper information not only for itself, but for all its dependencies (including transitives) as well. Thus, in
+contrast to the [POC Nixpkgs PR](https://github.com/NixOS/nixpkgs/pull/85103)
+and the [original design of the
+RFC](https://github.com/doronbehar/rfcs/blob/60d3825fdd4e6574b7e5d70264445d1c801368c6/rfcs/0075-declarative-wrappers.md#L251),
+prior to [the 1st
+meeting](https://github.com/NixOS/rfcs/pull/75#issuecomment-760942876),
+traversing all the inputs and the inputs' inputs, will not happen during eval
+time and only partly, during build time - every package already built will
+provide its reverse dependencies all the information they need about
+environment variables.
+
+Most of the work to do will be:
+
+1. Gather information about what environment variables are "linked" to each
+   package, and edit these derivations to include a `wrappers.json` in them.
+   This should be done with `makeWrapperAuto` as well, see (2).
+2. Design the `makeWrapperAuto` shell hook:
+  - It should introduce a shell function (to be called `combineWrappersInfo`) that
+    will allow piping a JSON string from `builtins.toJSON` and spit a
+    `wrappers.json` that will include both what was piped into it, and the
+    content from the package's various inputs' `wrappers.json` files.
+  - It should wrap the executables in `$out/bin/` according to
```

Comment:
Maybe we should also not wrap everything in `$out/bin` indiscriminately. A variable could specify which programs to wrap (as a list of glob patterns?), defaulting to `$out/bin/*` but permitting an override to _not_ wrap everything in `$out/bin` and/or _also_ wrap things outside `$out/bin`.

Also, should the wrapping apply to all outputs in multiple-output derivations? My guess (though not well-thought-out) is that `$out/bin/*`, `$bin/bin/*`, and `$out/libexec/**/*` should be wrapped by default.

Reply:
A list of glob patterns is a good idea.